### PR TITLE
add pytorch backend - se3 loss and gradient

### PR DESCRIPTION
**Pull request by johmathe**

### CODE REVIEW

File: geomstats/backend/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
+"""Pytorch based computation backend."""
+
+import torch
+import numpy as np
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: import order

File: geomstats/backend/pytorch.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+def to_ndarray(x, to_ndim, axis=0):
+    if x.dim() == to_ndim - 1:
+        x = torch.unsqueeze(x, dim=axis)
+    assert x.dim() >= to_ndim
+    return x
+
+
+def sqrt(val):
+    return torch.sqrt(torch.tensor(val).double())
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: why .double()?

File: geomstats/backend/numpy_linalg.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def matrix_rank(*args, **kwargs):
 
 
 def eigvalsh(*args, **kwargs):
-    return np.linalg.eigvalsh(*args, **kwargs)
+    return torch.eig(*args, **kwargs)
```
Comment (ninamiolane) [inline geomstats/backend/numpy_linalg.py]: this is in numpy_linalg?

File: geomstats/backend/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+
+def norm(val, axis):
+    return torch.linalg.norm(val, axis=axis)
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: do we want to create pytorch_linalg.py?

File: geomstats/backend/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+
+def rand(*args, **largs):
+    return torch.random.rand(*args, **largs)
+
+
+def isclose(*args, **kwargs):
+    print(array(args[1]))
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: left-over?

File: geomstats/backend/pytorch.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+def equal(*args, **kwargs):
+    return torch.equal(*args, **kwargs)
+
+
+def floor(*args, **kwargs):
+    return torch.floor(*args, **kwargs)
+
+
+def cross(*args, **kwargs):
+    print(args[1])
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: leftover?

File: geomstats/backend/pytorch.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    return torch.triu_indices(*args, **kwargs)
+
+
+def where(*args, **kwargs):
+    return torch.where(*args, **kwargs)
+
+
+def tile(*args, **kwargs):
+    # TODO(johmathe): Native tile implementation
+    print(np.tile(*args, **kwargs))
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: do you want to keep these prints?

File: geomstats/backend/torch.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
+"""Tensorflow based computation backend."""
```
Comment (ninamiolane) [inline geomstats/backend/torch.py]: Why is it all tensorflow herE?

Comment (ninamiolane) [reply geomstats/backend/torch.py]: (and file is named torch)

File: geomstats/invariant_metric.py [hunk 9]
```diff
@@ -31,11 +31,11 @@ def __init__(self, group,
         assert mat_shape == (1,) + (group.dimension, ) * 2, mat_shape
 
         assert left_or_right in ('left', 'right')
-
-        eigenvalues = gs.linalg.eigvalsh(inner_product_mat_at_identity)
-        n_pos_eigval = gs.sum(eigenvalues > 0)
-        n_neg_eigval = gs.sum(eigenvalues < 0)
-        n_null_eigval = gs.sum(eigenvalues == 0)
+        eigenvalues = gs.linalg.eigvalsh(inner_product_mat_at_identity[0])
```
Comment (ninamiolane) [inline geomstats/invariant_metric.py]: are you sure about the [0]? no other tests break?

File: geomstats/special_orthogonal_group.py [hunk 10]
```diff
@@ -25,11 +25,10 @@ def closest_rotation_matrix(mat):
     assert mat_dim_1 == mat_dim_2
 
     if mat_dim_1 == 3:
-        mat_unitary_u, diag_s, mat_unitary_v = gs.linalg.svd(mat)
+        mat_unitary_u, diag_s, mat_unitary_v = gs.linalg.svd(mat[0])
```
Comment (ninamiolane) [inline geomstats/special_orthogonal_group.py]: same question

File: geomstats/special_orthogonal_group.py [hunk 11]
```diff
@@ -150,11 +149,11 @@ def regularize(self, point):
 
         if vec_dim == 3:
             angle = gs.linalg.norm(point, axis=1)
-            regularized_point = point.astype('float64')
+            regularized_point = point
```
Comment (ninamiolane) [inline geomstats/special_orthogonal_group.py]: cool!

File: examples/gradient_descent_s2.py [hunk 12]
```diff
@@ -44,7 +48,7 @@ def gradient_descent(start,
         tangent_vec = manifold.projection_to_tangent_space(
                 vector=euclidean_grad, base_point=x)
         x = manifold.metric.exp(base_point=x, tangent_vec=tangent_vec)[0]
-        if np.abs(loss(x) - loss(x_prev)) <= precision:
+        if np.abs(loss(x, use_gs=True) - loss(x_prev, use_gs=True)) <= precision:
```
Comment (ninamiolane) [inline examples/gradient_descent_s2.py]: why this?

File: examples/plot_hexagon_h2.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
+"""
+Plot a grid on H2
```
Comment (ninamiolane) [inline examples/plot_hexagon_h2.py]: update docstring

File: examples/gradient_descent_s2.py [hunk 14]
```diff
@@ -85,16 +89,21 @@ def generate_well_behaved_matrix():
...
+    gs.seed(1985)
     A = generate_well_behaved_matrix()
-    loss = lambda x: np.matmul(x.T, np.matmul(A, x))  # NOQA
-    grad = lambda x: 2 * np.matmul(A, x)  # NOQA
-    initial_point = np.array([0., 1., 0.])
+
+    def grad(x):
+        return 2 * gs.matmul(A, x)
+
+    def loss(x, use_gs=False):
```
Comment (ninamiolane) [inline examples/gradient_descent_s2.py]: why is use_gs needed?

File: geomstats/backend/pytorch.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
+"""Pytorch based computation backend."""
+
+import numpy as np
+import torch
+
+int8 = 'torch.ByteTensor'
+int32 = 'torch.LongTensor'
+double = 'torch.DoubleTensor'
+float64 = 'torch.DoubleTensor'
+float16 = 'torch.Float'
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: NIT: float16 before float64?

File: geomstats/backend/pytorch.py [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+def shape(val):
+    return val.shape
+
+
+def dot(a, b):
+    return torch.from_numpy(np.dot(a, b))
+
+
+def maximum(a, b):
+    return torch.max(array(a), array(b))
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: why array(a) and array(b)?

File: geomstats/backend/pytorch.py [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+def matmul(*args, **kwargs):
+    return torch.matmul(*args, **kwargs)
+
+
+def sum(*args, **kwargs):
+    return torch.sum(*args, **kwargs)
+
+
+def einsum(*args, **kwargs):
+    return torch.from_numpy(np.einsum(*args, **kwargs)).float()
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: why casting to float?

File: geomstats/backend/pytorch.py [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+    if type(x) == np.ndarray:
+        x = torch.from_numpy(x)
+    if x.dim() == to_ndim - 1:
+        x = torch.unsqueeze(x, dim=axis)
+    assert x.dim() >= to_ndim
+    return x
+
+
+def sqrt(val):
+    return torch.sqrt(torch.tensor(val).float())
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: why do you need to convert val to a tensor here, and not elsewhere?

Comment (ninamiolane) [reply geomstats/backend/pytorch.py]: why casting in float?

File: geomstats/backend/pytorch_linalg.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    svds = np.linalg.svd(*args, **kwargs)
+    return torch.from_numpy(svds[0]), torch.from_numpy(svds[1]), torch.from_numpy(svds[2])
+
+
+def det(*args, **kwargs):
+    return torch.from_numpy(np.linalg.det(*args, **kwargs))
+
+
+def norm(x, ord=2, axis=None, keepdims=False):
+    assert keepdims is False
```
Comment (ninamiolane) [inline geomstats/backend/pytorch_linalg.py]: what is this?

File: geomstats/backend/pytorch_random.py [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+def rand(*args, **kwargs):
+    return torch.rand(*args, **kwargs)
+
+
+def randint(*args, **kwargs):
+    return torch.randint(*args, **kwargs)
+
+
+def seed(*args, **kwargs):
+    return torch.manual_seed(*args, **kwargs)
```
Comment (ninamiolane) [inline geomstats/backend/pytorch_random.py]: on geomstats backend: we consistently try to be as close to numpy API everywhere, right?

File: geomstats/hypersphere.py [hunk 21]
```diff
@@ -184,7 +182,7 @@ def log(self, point, base_point):
         angle = gs.arccos(cos_angle)
 
         mask_0 = gs.isclose(angle, 0.0)
-        mask_else = gs.equal(mask_0, False)
+        mask_else = gs.equal(mask_0, gs.cast(gs.array(False), gs.int8))
```
Comment (ninamiolane) [inline geomstats/hypersphere.py]: why?

File: geomstats/invariant_metric.py [hunk 22]
```diff
@@ -31,11 +31,10 @@ def __init__(self, group,
...
 
         assert left_or_right in ('left', 'right')
-
         eigenvalues = gs.linalg.eigvalsh(inner_product_mat_at_identity)
-        n_pos_eigval = gs.sum(eigenvalues > 0)
-        n_neg_eigval = gs.sum(eigenvalues < 0)
-        n_null_eigval = gs.sum(eigenvalues == 0)
+        n_pos_eigval = gs.sum(eigenvalues > 0.0)
+        n_neg_eigval = gs.sum(eigenvalues < 0.0)
+        n_null_eigval = gs.sum(eigenvalues == 0.0)
```
Comment (ninamiolane) [inline geomstats/invariant_metric.py]: why?

File: geomstats/special_orthogonal_group.py [hunk 23]
```diff
@@ -27,9 +27,8 @@ def closest_rotation_matrix(mat):
     if mat_dim_1 == 3:
         mat_unitary_u, diag_s, mat_unitary_v = gs.linalg.svd(mat)
         rot_mat = gs.matmul(mat_unitary_u, mat_unitary_v)
-
-        mask = gs.where(gs.linalg.det(rot_mat) < 0)
-        new_mat_diag_s = gs.tile(gs.diag([1, 1, -1]), len(mask))
+        mask = gs.nonzero(gs.linalg.det(rot_mat) < 0)
+        new_mat_diag_s = gs.tile(gs.diag(gs.array([1, 1, -1])), len(mask))
```
Comment (ninamiolane) [inline geomstats/special_orthogonal_group.py]: why?

File: geomstats/backend/pytorch_random.py (hunk 20 above)
Comment (johmathe) [reply geomstats/backend/pytorch_random.py]: yes good tach

File: examples/gradient_descent_s2.py (hunk 14 above)
Comment (johmathe) [reply examples/gradient_descent_s2.py]: for rendering

File: geomstats/backend/pytorch.py (hunk 16 above)
Comment (johmathe) [reply geomstats/backend/pytorch.py]: if not cast to an array it will fail trying to compare values

File: geomstats/backend/pytorch_linalg.py (hunk 19 above)
Comment (johmathe) [reply geomstats/backend/pytorch_linalg.py]: leftover debug - good catch
